internal/proxyservice/http/proxy_service.go

```diff
@@ -12,18 +19,269 @@ import (
+		logger.Warnf(msg, err)
+		caCertPool = x509.NewCertPool()
+	}
+
+	transport := &gohttp.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs: caCertPool,
+		},
+	}
+
+	return &proxyService{
+		transport:   transport,
+		subservices: subservices,
+		listener:    sharedListener,
+		logger:      logger,
+		done:        false,
+	}, nil
+}
+
+type proxyService struct {
+	transport   *gohttp.Transport
+	done        bool
+	listener    net.Listener
+	logger      log.Logger
+	subservices []Subservice
+}
+
+func (proxy *proxyService) matchingSubservices(r *gohttp.Request) []Subservice {
+	var matchingSubs []Subservice
+	for _, sub := range proxy.subservices {
+		if sub.Matches(r.URL.String()) {
+			matchingSubs = append(matchingSubs, sub)
+		}
+	}
+	return matchingSubs
+}
+
+// selectSubservice finds a subservice matching the request, and issues warnings
+// when there are no matches or more than one match.
+func (proxy *proxyService) selectSubservice(r *gohttp.Request) *Subservice {
+	matchingSubs := proxy.matchingSubservices(r)
+
+	// No match: Warn!
+	if len(matchingSubs) == 0 {
+		msg := "No subservices matched request '%s'"
+		proxy.logger.Warnf(msg, r.URL.String(), matchingSubs)
+		return nil
+	}
+
+	// Multiple matches: Warn!
+	if len(matchingSubs) > 1 {
+		msg := "Multiple subservices matched request '%s': %v\n"
+		proxy.logger.Warnf(msg, r.URL.String(), matchingSubs)
+	}
+
+	// Select first (or only) match
+	subservice := matchingSubs[0]
+	msg := "Using connector '%s' for request %s"
+	proxy.logger.Debugf(msg, subservice.ConnectorID, r.URL.String())
+	return &subservice
+}
+
+
+// ServeHTTP exists to implement the go_http.Handler interface
+func (proxy *proxyService) ServeHTTP(w gohttp.ResponseWriter, r *gohttp.Request) {
+	logger := proxy.logger
+
+	// Log request
+
+	logMsg := "Got request %v %v %v %v"
+	logger.Infoln(logMsg, r.URL.Path, r.Host, r.Method, r.URL.String())
+
+	// Validate request
+
+	if !proxy.validateProxyServerRules(w, r) {
+		return
+	}
+
+	// Remove headers intended only for proxy server.
+
+	r.Header.Del("Proxy-Connection")
+	r.Header.Del("Proxy-Authenticate")
+	r.Header.Del("Proxy-Authorization")
+
+	// Select a subservice
+
+	subservice := proxy.selectSubservice(r)
+
+	// No match: Send request with no authentication.
+	if subservice == nil {
+		proxy.handleRequest(w, r)
+		return
+	}
+
+	// Get current credential values
+
+	creds, err := subservice.RetrieveCredentials()
+	defer internal.ZeroizeCredentials(creds)
+
+	if err != nil {
+		gohttp.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Authenticate request
+
+	err = subservice.Authenticate(r, creds)
+	if err != nil {
+		gohttp.Error(w, err.Error(), 500)
+		return
+	}
+
+	// Send request to target service
+
+	proxy.handleRequest(w, r)
+}
+
+// validateProxyServerRules ensures that the request being made is a valid
+// request for a proxy server.
+func (proxy *proxyService) validateProxyServerRules(
```

Comment:
the naming seems a bit dishonest here. i might be mis-assigning meaning to the term "validation". my expectation, which might be wrong, is for validation to tell me if the input is valid or not, without side-effects.

Reply:
Yeah I think that is a mis-assignment, at least in my mind.

Following the "verb phrase for side effect" rule `ValidateXX` implies that it will do a validation and (in other langs) throw errors or (in Go) return an error or (in this case) write to a log.  If you were writing a method that just returned a boolean it's name would be of the form `IsValidXXX`.  You can see I followed this convention here: 

https://github.com/cyberark/secretless-broker/blob/master/internal/profile/profile.go#L37 

Ofc this isn't set in stone but the distinction is useful and the naming convention is in line with the more general one so I like it.